## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [4]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

FileNotFoundError: [Errno 2] No such file or directory: 'Resources/client_dataset.csv'

In [ ]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [ ]:
# Use the describe function to gather some basic statistics
basic_stats = df.describe()
print(basic_stats)

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [ ]:
# Use this space to do any additional research
# and familiarize yourself with the data.

missing_values = df.isnull().sum()
print(missing_values)

data_types = df.dtypes
print(data_types)

unique_values = df.nunique()
print(unique_values)

import matplotlib.pyplot as plt
df.hist(bins=50, figsize=(20, 15))
plt.show()

correlation_matrix = df.corr()
print(correlation_matrix)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [ ]:
# What three item categories had the most entries?
# Find the three item categories with the most entries
top_categories = df['category'].value_counts().head(3)
print(top_categories)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [ ]:
# For the category with the most entries, which subcategory had the most entries?
# Filter the DataFrame to only include rows with the most common category
consumables_df = df[df['category'] == 'consumables']

# Find the subcategory with the most entries in the filtered DataFrame
top_subcategory = consumables_df['subcategory'].value_counts().head(1)
print(top_subcategory)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [ ]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
print(top_clients)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [ ]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = top_clients.index.tolist()
print(top_client_ids)


[33615, 66037, 46820, 38378, 24741]

In [ ]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = top_clients.index[0]

top_client_df = df[df['client_id'] == top_client_id]

total_units_ordered = top_client_df['qty'].sum()
print(total_units_ordered)


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [ ]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
# Create a column that calculates the subtotal for each line
df['subtotal'] = df['unit_price'] * df['qty']
print(df.head())  # Display the first few rows to check the new column


unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [ ]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
# Create a function to calculate the shipping price
def calculate_shipping(unit_weight):
    if unit_weight > 50:
        return unit_weight * 7
    else:
        return unit_weight * 10

# Apply the function to create a new column for the shipping price
df['shipping_price'] = df['unit_weight'].apply(calculate_shipping)
print(df.head())  # Display the first few rows to check the new column


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [ ]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

# Calculate the subtotal
df['subtotal'] = df['unit_price'] * df['qty']

# Create a function to calculate the shipping price
def calculate_shipping(unit_weight):
    if unit_weight > 50:
        return unit_weight * 7
    else:
        return unit_weight * 10

# Apply the function to create a new column for the shipping price
df['shipping_price'] = df['unit_weight'].apply(calculate_shipping)

# Calculate the total price
sales_tax_rate = 0.0925
df['total_price'] = df['subtotal'] + df['shipping_price'] + (df['subtotal'] * sales_tax_rate)

print(df.head())  # Display the first few rows to check the new column


line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [ ]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
# Create a column for the line cost
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
print(df.head())  # Display the first few rows to check the new column


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [ ]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - df['line_cost']
print(df.head())  # Display the first few rows to check the new column


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [ ]:
# Check your work using the totals above
# Provided total prices for validation
provided_totals = {
    2742071: 152811.89,
    2173913: 162388.71,
    61289929: 923441.25
}

# Compare the calculated total prices with the provided totals
for order_id, provided_total in provided_totals.items():
    calculated_total = df[df['order_id'] == order_id]['total_price'].sum()
    print(f"Order ID {order_id}:")
    print(f"Calculated Total: ${calculated_total:.2f}")
    print(f"Provided Total: ${provided_total}")
    print(f"Match: {calculated_total == provided_total}\n")


Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [ ]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# Assuming top_client_ids contains the IDs of the top 5 clients
top_client_ids = [id1, id2, id3, id4, id5]  # Replace with actual client IDs

# Initialize a dictionary to store the total spending for each top client
client_spending = {}

# Calculate the total spending for each top client
for client_id in top_client_ids:
    total_spent = df[df['client_id'] == client_id]['total_price'].sum()
    client_spending[client_id] = total_spent

# Display the results
for client_id, total_spent in client_spending.items():
    print(f"Client ID {client_id} spent a total of ${total_spent:.2f}")



33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [ ]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
# Initialize a dictionary to store the summary information for each top client
summary_data = {
    'client_id': [],
    'total_units_purchased': [],
    'total_shipping_price': [],
    'total_revenue': [],
    'total_profit': []
}

# Calculate the totals for each top client
for client_id in top_client_ids:
    client_df = df[df['client_id'] == client_id]
    summary_data['client_id'].append(client_id)
    summary_data['total_units_purchased'].append(client_df['qty'].sum())
    summary_data['total_shipping_price'].append(client_df['shipping_price'].sum())
    summary_data['total_revenue'].append(client_df['total_price'].sum())
    summary_data['total_profit'].append(client_df['profit'].sum())

# Create the summary DataFrame
summary_df = pd.DataFrame(summary_data)
print(summary_df)


client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [ ]:
# Format the data and rename the columns to names suitable for presentation.
summary_df.rename(columns={
    'client_id': 'Client ID',
    'total_units_purchased': 'Total Units Purchased',
    'total_shipping_price': 'Total Shipping Price',
    'total_revenue': 'Total Revenue',
    'total_profit': 'Total Profit'
}, inplace=True)

summary_df['Total Units Purchased'] = summary_df['Total Units Purchased'].astype(int)
summary_df['Total Shipping Price'] = summary_df['Total Shipping Price'].map('${:,.2f}'.format)
summary_df['Total Revenue'] = summary_df['Total Revenue'].map('${:,.2f}'.format)
summary_df['Total Profit'] = summary_df['Total Profit'].map('${:,.2f}'.format)

print(summary_df)

# Define the money columns. 
money_columns = ['Total Shipping Price', 'Total Revenue', 'Total Profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return f"${x / 1_000_000:,.2f}M"

# Apply the currency_format_millions function to only the money columns. 
# Apply the currency_format_millions function to only the money columns
for column in money_columns:
    summary_df[column] = summary_df[column].apply(lambda x: currency_format_millions(float(x.replace('$','').replace(',',''))))

# Rename the columns to reflect the change in the money format. 
# Rename the columns to reflect the change in the money format
summary_df.rename(columns={
    'Total Shipping Price': 'Total Shipping Price (Millions)',
    'Total Revenue': 'Total Revenue (Millions)',
    'Total Profit': 'Total Profit (Millions)'
}, inplace=True)

print(summary_df)


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
# Sort the data by "Total Profit (Millions)" from highest to lowest
sorted_summary_df = summary_df.sort_values(by='Total Profit (Millions)', ascending=False)

print(sorted_summary_df)


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995